pilot/pkg/status/resourcelock.go

```diff
@@ -14,47 +14,168 @@
+	Push(target Resource, progress Progress)
+	// Run the loop until a signal on the context
+	Run(ctx context.Context)
+	// Delete a task
+	Delete(target lockResource)
 }
 
-func (r *ResourceLock) Lock(i interface{}) {
-	r.retrieveMutex(i).Lock()
+type queueImpl struct {
+	delay   time.Duration
+	tasks   []lockResource
+	cond    *sync.Cond
+	closing bool
+	work    Task
+
+	cacheLock   sync.Mutex
+	cache       map[lockResource]*cacheEntry
+	workerCount int
+	maxWorkers  int
 }
 
-func (r *ResourceLock) Unlock(i interface{}) {
-	r.retrieveMutex(i).Unlock()
+// NewQueue instantiates a queue with a processing function
+func NewQueue(errorDelay time.Duration, work Task, maxWorkers int) WorkerQueue {
+	return &queueImpl{
+		delay:       errorDelay,
+		tasks:       make([]lockResource, 0),
+		closing:     false,
+		cond:        sync.NewCond(&sync.Mutex{}),
+		work:        work,
+		workerCount: 0,
+		maxWorkers:  maxWorkers,
+		cache:       make(map[lockResource]*cacheEntry),
+	}
 }
 
-// returns value indicating if init was necessary
-func (r *ResourceLock) init() bool {
-	if r.listing == nil {
-		r.masterLock.Lock()
-		defer r.masterLock.Unlock()
-		// double check, per pattern
-		if r.listing == nil {
-			r.listing = make(map[interface{}]*sync.Mutex)
-		}
-		return true
+func (q *queueImpl) Push(target Resource, progress Progress) {
+	q.cacheLock.Lock()
+	defer q.cacheLock.Unlock()
+	key := convert(target)
+	_, inqueue := q.cache[key]
+	q.cache[key] = &cacheEntry{
+		cacheVal:      &target,
+		cacheProgress: &progress,
+	}
+	if !inqueue {
+		q.enqueue(key)
 	}
-	return false
+	q.maybeAddWorker()
 }
 
-func (r *ResourceLock) retrieveMutex(i interface{}) *sync.Mutex {
-	if !r.init() {
-		r.masterLock.RLock()
-		if result, ok := r.listing[i]; ok {
-			r.masterLock.RUnlock()
-			return result
+func (q *queueImpl) Delete(target lockResource) {
+	q.cacheLock.Lock()
+	defer q.cacheLock.Unlock()
+	delete(q.cache, target)
+}
+
+func (q *queueImpl) enqueue(item lockResource) {
+	q.cond.L.Lock()
+	defer q.cond.L.Unlock()
+	if !q.closing {
+		q.tasks = append(q.tasks, item)
+	}
+	q.cond.Signal()
+}
+
+func (q *queueImpl) maybeAddWorker() {
+	q.cond.L.Lock()
+	if q.workerCount >= q.maxWorkers {
+		q.cond.L.Unlock()
+		return
+	}
+	q.workerCount++
+	q.cond.L.Unlock()
+	go func() {
+		for {
+			q.cond.L.Lock()
+			for !q.closing && len(q.tasks) == 0 {
+				if q.workerCount > 1 {
+					q.workerCount--
+					q.cond.L.Unlock()
+					return
+				}
+				q.cond.Wait()
+			}
+
+			if len(q.tasks) == 0 {
+				q.cond.L.Unlock()
+				// We must be shutting down.
+				return
+			}
+
+			var target lockResource
+			target, q.tasks = q.tasks[0], q.tasks[1:]
+			q.cond.L.Unlock()
+
+			var c *cacheEntry
+			q.cacheLock.Lock()
+			c, ok := q.cache[target]
+			if !ok {
+				// this element has been deleted, move along
+				q.cacheLock.Unlock()
+				continue
+			} else {
+				delete(q.cache, target)
+			}
+			q.cacheLock.Unlock()
+
+			if err := q.work(c); err != nil {
```

Comment:
nit: in our current usage, `work` always returns nil. I would recommend making it not allow an error, then we don't have this un-tested retrying code - if we need it later we can always add it (along with test). Also cleans up `errorDelay`